apps/external_ledger_db/lib/external_ledger_db/wallet.ex

```diff
@@ -0,0 +1,174 @@
+# Copyright 2017-2019 OmiseGO Pte Ltd
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+defmodule ExternalLedgerDB.Wallet do
+  @moduledoc """
+  Ecto Schema representing external ledger wallets. A wallet is made up of
+  a unique address and the ID associated with it.
+  """
+  use Ecto.Schema
+  import Ecto.{Changeset, Query}
+  alias ExternalLedgerDB.Repo
+
+  @behaviour Utils.Ledgers.WalletSchema
+
+  @primary_key {:uuid, Ecto.UUID, autogenerate: true}
+  @timestamps_opts [type: :naive_datetime_usec]
+
+  @hot "hot"
+  @cold "cold"
+
+  def hot, do: @hot
+  def cold, do: @cold
+
+  @ethereum "ethereum"
+  @omg_network "omg_network"
+
+  def ethereum, do: @ethereum
+  def omg_network, do: @omg_network
+
+  schema "wallet" do
+    field(:address, :string)
+    field(:adapter, :string)
+    field(:primary, :boolean, default: false)
+    field(:type, :string)
+    field(:public_key, :string)
+    field(:encrypted_private_key, ExternalLedgerDB.Encrypted.String)
+    field(:metadata, :map, default: %{})
+    field(:encrypted_metadata, ExternalLedgerDB.Encrypted.Map, default: %{})
+
+    timestamps()
+  end
+
+  @doc """
+  Validate the wallet attributes.
+  """
+  @spec changeset(%__MODULE__{}, map()) :: Ecto.Changeset.t()
+  def changeset(%__MODULE__{} = wallet, attrs) do
+    wallet
+    |> cast(attrs, [
+      :address,
+      :adapter,
+      :primary,
+      :type,
+      :public_key,
+      :encrypted_private_key,
+      :metadata,
+      :encrypted_metadata
+    ])
+    |> validate_required([
+      :address,
+      :adapter,
+      :primary,
+      :type,
+      :public_key,
+      :metadata,
+      :encrypted_metadata
+    ])
+    |> validate_inclusion(:adapter, [@ethereum, @omg_network])
```

Comment:
Ethereum and the OMG network are not exclusive, they work together. The same wallet (same Ethereum address) can have funds on Ethereum AND Plasma at the same time. This should validate the inclusion within the registered list of adapters instead. We'll later have an adapter that will basically be `ethereum + omg_network` (but we could also have an adapter just for `ethereum`, which will be the case at first).

Reply:
A bit confused. I was thinking that each `ExternalLedgerDB.Wallet` should belong to one adapter…

And the same wallet with same address on ethereum and omg_plasma would be considered two `ExternalLedgerDB.Wallet`, because technically they share nothing apart from the same pub/priv keys? (which may not be true too for other chains). The system doesn’t need to know that they are related?

Also… do you mean a wallet.adapter should be an array like `wallet.adapter == ["ethereum", "omg_network"]`? If so I feel there’ll be a lot of conditioning based on that array compared to 1 wallet 1 adapter.